### PR TITLE
docs(frontends): propose a volume-based "frontend"

### PR DESCRIPTION
**Pull request by silasbw**
Propose and document an API for writing secret data to volumes.

See: https://github.com/godaddy/kubernetes-external-secrets/issues/78

### CODE REVIEW

File: api.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  generateName: pod-example-
+  annotations:
+    externalsecrets.kubernetes-client.io/volumes: >
+      [
+        {"name": "db-secrets", "externalSecretName": "db-secrets"},
```
Comment (silasbw) [inline api.md]: Another way I was thinking about this was to mirror volume declarations.  E.g.,

```json
{"name": "db-secrets", "externalSecret": { "externalSecretName": "db-secrets" } }
```

https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod

I thought that my be too verbose.  On the other hand it parallels an existing Kubernetes pattern.

Comment (JacopoDaeli) [reply api.md]: Maybe 
```
{"name": "db-secrets", "externalSecret": { "name": "db-secrets" } }
```
?

Comment (jeffpearce) [reply api.md]: I like that even better than externalSecretName

Comment (JacopoDaeli) [reply api.md]: Yeah, it makes more sense.

Comment (silasbw) [reply api.md]: I also like "name" better than "externalSecretName", but it departs from the kubernetes convention (see the link I included above), so I'd encourage us to follow the convention of go with "externalSecretName".

**Comment by silasbw** [comment]
@JacopoDaeli see https://github.com/godaddy/kubernetes-external-secrets/issues/46#issuecomment-498426782

**Comment by silasbw** [comment]
@max-lobur do you have time to give this proposal a quick look?

File: api.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+`externalsecrets.kubernetes-client.io/volumes` annotation to a `Pod`
+manifest:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  generateName: pod-example-
+  annotations:
+    externalsecrets.kubernetes-client.io/volumes: >
```
Comment (max-lobur) [inline api.md]: should be possible to write in yaml?
```
externalsecrets.kubernetes-client.io/volumes:
- name: db-secrets
  externalSecret:
    externalSecretName: db-secrets
...
```

**Comment by max-lobur** [comment]
The last suggestion breaks the frontend abstraction tho

File: api.md (hunk 2 above)
Comment (silasbw) [reply api.md]: Doh, yeah, updated.

**Comment by silasbw** [comment]
Thanks @max-lobur !  Kubernetes External Secrets will create a file in the volume for each element in the `.secretDescriptor.data` array of the `ExternalSecret` specified in the annotation.  We're trying to minor the behavior of the native `Secret` object.  For example:

```yml
apiVersion: v1
kind: Pod
metadata:
  generateName: pod-example-
  annotations:
    externalsecrets.kubernetes-client.io/volumes: |
      - name: "hello-service"
        externalSecret:
          externalSecretName: "hello-service"
spec:
  containers:
  - image: busybox
    name: busybox
    volumeMounts:
    - mountPath: /secrets
      name: hello-service
  volumes:
  - name: hello-service
    emptyDir:
      medium: Memory
---
apiVersion: 'kubernetes-client.io/v1'
kind: ExternalSecret
metadata:
  name: hello-service
secretDescriptor:
  backendType: secretsManager
  data:
    - key: hello-service/password
      name: password
    - key: hello-service/apikey
      name: apikey
```

would cause the Kubernetes External Secrets controller to create two file in `/secrets`, one named `password` and one named `apikey`.  Kubernetes External Secrets would write the contents of `hello-service/password` from AWS SM into `password` and the contents of `hello-service/apikey` from AWS SM into `apikey`.

File: api.md [hunk 3]
```diff
@@ -45,10 +62,32 @@ spec:
...
+  backendType: secretsManager
+  data:
+    - key: api/key
+      name: key
 ```
 
-The value of `externalsecrets.kubernetes-client.io/volumes` is a JSON
-array of volume configuration objects:
+The value of `externalsecrets.kubernetes-client.io/volumes` is a JSON or
+YAML serizlied array of volume configuration objects:
```
Comment (jeffpearce) [inline api.md]: serialized

File: api.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+* fetch the value of api/key from AWS Secrets Manager and write that
+  values to the api file in the client-secrets volume.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  generateName: pod-example-
+  annotations:
+    externalsecrets.kubernetes-client.io/volumes: |
```
Comment (max-lobur) [inline api.md]: Just curious, does the annotation has to be a multiline string `|`?
What if you just put a regular yaml there like
```
externalsecrets.kubernetes-client.io/volumes:
     - name: "db-secrets"
        externalSecret:
          externalSecretName: "db-secrets"
      - name: "client-secrets"
        externalSecret:
          externalSecretName: "client-secrets"
```
and then read it as array, without de-serealizing

Comment (silasbw) [reply api.md]: Kubernetes requires annotation values to be strings.

**Comment by silasbw** [comment]
Applications could still leverage environment variables but you'd probably need to set some application code to do that instead of using `secretRef`.  *E.g.*, initContainer writes values to volume, containers source them before starting the application process (`source /some/volume; ./server.sh`).

Definitely open to better ways of doing this.  More ambitious would be adding a `volumeRef` to the API (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#envvarsource-v1-core)

There's definitely some trade-offs when moving off of etcd / `Secrets` to the volume-based frontend.  It's something folks have been asking for and in some cases believe it's a worthwhile improvement in privacy and integrity.

**Comment by dudicoco** [comment]
Thanks for the reply @silasbw. 
I'm aware I can source environment variables beforehand, however, that wouldn't be very clean when used with third party applications since the Docker entrypoint will have to be replaced.

bank-vault has an interesting solution of injecting into the container's entrypoint:
https://banzaicloud.com/blog/inject-secrets-into-pods-vault-revisited/

"The init-container also changes the command of the container to run vault-env, instead of running the application directly"

Personally i'm leaning more towards using SSM along with native kubernetes secrets, but using the external-secrets controller means granting the pod access to all secrets (or permissions to assume role to all roles which can assume secrets) which is not ideal.

I think a good solution would be to bring up a job which uses the same IAM role as the deployment/statefulset/daemonset, the job will fetch the secret from Parameter Store and will update the appropriate secret in k8s.

Advantages:
1. Native k8s secrets
2. Granular access to Parameter Store.
3. One API call per deployment/statefulset/daemonset, less likely to throttle SSM API calls.
4. Can run as a cron job to keep secrets in sync with Parameter Store.

Let me know what you think!

**Comment by silasbw** [comment]
Thanks for the bank-vault link. Not sure if that will work for `ExternalSecrets`or not, but it might help inspire a better solution.

I agree we should have more fine-grained roles on the controller.  That's something we should do regardless of how the secret data is communicated (`Secret` vs `Volume`) to the running containers.  FWIW, we are going to support the `Volume`-based approach as an alternative to the `Secret` one (not as a replacement).

For getting finer-grained roles, would it work well for you use case if we supported specifying a name space and a role for the controller to watch and assume?

**Comment by dudicoco** [comment]
Do you have an estimation of when the volume based approach going to be implemented?

I think a nice alternative solution should give not take into account assuming a role, considering that it should be now possible to assign an IAM role directly to a pod's service account:
https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/

**Pull request merged**